drivers/tee/optee/optee.c

```diff
@@ -0,0 +1,500 @@
+	if (ret) {
+		LOG_ERR("Unable to free shared memory");
+	}
+
+	return (rc) ? rc : ret;
+}
+
+static int optee_cancel(const struct device *dev, uint32_t session_id, uint32_t cancel_id)
+{
+	int rc;
+	struct tee_shm *shm;
+	struct optee_msg_arg *marg;
+
+	rc = tee_add_shm(dev, NULL, OPTEE_MSG_NONCONTIG_PAGE_SIZE,
+			 OPTEE_MSG_GET_ARG_SIZE(0),
+			 TEE_SHM_ALLOC, &shm);
+	if (rc) {
+		LOG_ERR("Unable to get shared memory, rc = %d", rc);
+		return rc;
+	}
+
+	marg = shm->addr;
+	marg->num_params = 0;
+	marg->cmd = OPTEE_MSG_CMD_CANCEL;
+	marg->cancel_id = cancel_id;
+	marg->session = session_id;
+
+	rc = optee_call(dev, marg);
+
+	if (tee_rm_shm(dev, shm)) {
+		LOG_ERR("Unable to free shared memory");
+	}
+
+	return rc;
+}
+
+static int optee_invoke_func(const struct device *dev, struct tee_invoke_func_arg *arg,
+			     unsigned int num_param, struct tee_param *param)
+{
+	int rc, ret;
+	struct tee_shm *shm;
+	struct optee_msg_arg *marg;
+
+	if (!arg) {
+		return -EINVAL;
+	}
+
+	rc = tee_add_shm(dev, NULL, OPTEE_MSG_NONCONTIG_PAGE_SIZE,
+			 OPTEE_MSG_GET_ARG_SIZE(num_param),
+			 TEE_SHM_ALLOC, &shm);
+	if (rc) {
+		LOG_ERR("Unable to get shared memory, rc = %d", rc);
+		return rc;
+	}
+
+	marg = shm->addr;
+	memset(marg, 0, OPTEE_MSG_GET_ARG_SIZE(num_param));
+
+	marg->num_params = num_param;
+	marg->cmd = OPTEE_MSG_CMD_INVOKE_COMMAND;
+	marg->func = arg->func;
+	marg->session = arg->session;
+
+	rc = param_to_msg_param(param, num_param, marg->params);
+	if (rc) {
+		goto out;
+	}
+
+	arg->ret = optee_call(dev, marg);
+	if (arg->ret) {
+		arg->ret_origin = TEEC_ORIGIN_COMMS;
+		goto out;
+	}
+
+	rc = msg_param_to_param(param, num_param, marg->params);
+	if (rc) {
+		arg->ret = TEEC_ERROR_COMMUNICATION;
+		arg->ret_origin = TEEC_ORIGIN_COMMS;
+		goto out;
+	}
+
+	arg->ret = marg->ret;
+	arg->ret_origin = marg->ret_origin;
+out:
+	ret = tee_rm_shm(dev, shm);
+	if (ret) {
+		LOG_ERR("Unable to free shared memory");
+	}
+
+	return (rc) ? rc : ret;
+}
+
+static int optee_shm_register(const struct device *dev, void *addr, uint64_t size, uint32_t flags,
+			      struct tee_shm *shm)
+{
+	return 0;
+}
+
+static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+{
+	return 0;
+}
+
+static int optee_suppl_recv(const struct device *dev, uint32_t func, unsigned int num_params,
+			    struct tee_param *param)
+{
+	return 0;
+}
+
+static int optee_suppl_send(const struct device *dev, unsigned int num_params,
+			    struct tee_param *param)
+{
+	return 0;
+}
+
+static int set_optee_method(void)
+{
+	const char *method;
+
+	method = DT_PROP(DT_INST(0, DT_DRV_COMPAT), method);
+
+	if (!strcmp("hvc", method)) {
+		optee_data.smc_call = optee_smccc_hvc;
+	} else if (!strcmp("smc", method)) {
+		optee_data.smc_call = optee_smccc_smc;
+	} else {
+		LOG_ERR("Invalid smc_call method");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int optee_init(const struct device *dev)
+{
+	if (set_optee_method()) {
+		return -ENOTSUP;
+	}
+
```

Comment:
Linux kernel drivers performs basic communication testing by exchanging caps with OP-TEE. You also need do this.

Reply:
Created task in jira regrading this

Reply:
Could you please create the same task on GitHub? Also, please add TODO to the code

Reply:
Added TODO and created an Issue
https://github.com/xen-troops/zephyr/issues/61